File: NPC_SUB_DIGRAFH_PYTHON_CODE/Algoritmos.py
# Algoritmos.py
import networkx as nx
from itertools import combinations


# ------------FUNCION ENCARGADA DE CONTAR OPERACIONES----------------
class OpCounter:
    """Contador de operaciones (local o compartido)."""
    def __init__(self, shared = None):
        # Si no se pasa valor compartido usamos un int normal
        if shared is None:
            self._local = 0
            self._shared = None
        else:
            self._local = None
            self._shared = shared            # mp.Value('i', 0)

    def inc(self, n: int = 1) -> None:
        if self._shared is None:
            self._local += n
        else:                               # acceso con lock
            with self._shared.get_lock():
                self._shared.value += n

    @property
    def ops(self) -> int:
        return self._local if self._shared is None else self._shared.value

# ----------------------------------------------------------------------
# 1. Fuerza bruta
# ----------------------------------------------------------------------
def BipartiteSubsetBF(G : nx.Graph, k : int, counter: OpCounter | None = None) -> tuple[bool, int]:
    counter = counter or OpCounter()


    G1 = nx.Graph(); counter.inc()
    G1.add_nodes_from(G.nodes); counter.inc(len(G))  #O(V)
    if k < 0:
        raise ValueError("k debe ser mayor o igual a 0")
    
    if k > len(G.edges()):
        raise ValueError(f"k debe ser menor o igual al numero de aristas del grafo ({len(G.edges())})")
    
    if  k <= 2:
        counter.inc()
        return True , counter.ops


    for i in combinations(G.edges(), k): #Se repite maximo E (Ningun posible subgrafo de ese tamaño)
    #Peor Caso k = |E|/2
        G1.add_edges_from(i)
        counter.inc(len(i)) # O(E) 

        counter.inc()
        if nx.is_bipartite(G1): # O(V + E)
            return True, counter.ops
        
        else:
            G1.remove_edges_from(i)
            counter.inc(len(i))  #O(E)
    return False, counter.ops

# ----------------------------------------------------------------------
# 2. Divide & Conquer
# ----------------------------------------------------------------------
def BipartiteSubsetDC(G : nx.Graph, k : int,counter: OpCounter | None = None) -> tuple[bool, int]:
    counter = counter or OpCounter()

    if k < 0:
      raise ValueError("k debe ser mayor o igual a 0")
    if k > len(G.edges()):
        raise ValueError(f"k debe ser menor o igual al numero de aristas del grafo ({len(G.edges())})")
    __G = nx.Graph()
    __G.add_nodes_from(G.nodes)
    if k <= 2:
        return True, counter.ops
    
    def __BipartiteSubsetDC(_G : nx.Graph, k_left : int,edges = list) -> bool:
        if k_left == 0 :
            return True
        if len(edges)< k_left:
            return False
        
        G1 = _G.copy(); counter.inc()
        G2 = _G.copy(); counter.inc()

        G2.add_edge(edges[0][0],edges[0][1]); counter.inc()

        counter.inc()
        a = __BipartiteSubsetDC(G1,k_left,edges[1:])
        counter.inc()
        b = __BipartiteSubsetDC(G2,k_left-1,edges[1:]) if nx.is_bipartite(G2) else False
        
        if b:
            return True
        if a:
            return True

    return __BipartiteSubsetDC(__G,k,list(G.edges())), counter.ops


# ----------------------------------------------------------------------
# 3. Backtracking
# ----------------------------------------------------------------------
def BipartiteSubsetBT(G : nx.Graph, k : int,counter: OpCounter | None = None) -> tuple[bool, int]:
    counter = counter or OpCounter()

    if k < 0:
        raise ValueError("k debe ser mayor o igual a 0")
    if k > len(G.edges()):
        raise ValueError(f"k debe ser menor o igual al numero de aristas del grafo ({len(G.edges())})")
    if k <= 2 :
        return True, counter.ops
    
    edges_list = list(G.edges())

    def __BipartiteSubsetBT(_G : nx.Graph, k_pass : int =  0, init = 0) -> bool:

        if k == k_pass:
            return True
            
        if k - k_pass > len(edges_list) - init  : #Faltan Aristas
            return False
        
        for i in range(init, len(edges_list)):
            u,v = edges_list[i]
            _G.add_edge(u,v) ; counter.inc()# +1
            if nx.is_bipartite(_G): # O (V + E)
                counter.inc()                 # cuenta la llamada
                if __BipartiteSubsetBT(_G, k_pass + 1, i + 1):
                    return True
            _G.remove_edge(u,v); counter.inc() #+1
        return False

    G1 = nx.Graph()
    G1.add_nodes_from(G.nodes); counter.inc(len(G)) #O (V)
    edge_list = list(G.edges())
    return __BipartiteSubsetBT(G1),counter.ops

File: NPC_SUB_DIGRAFH_JAVA_CODE/src/CodigoJava/Algoritmos.java
package CodigoJava;

import org.jgrapht.Graph;
import org.jgrapht.GraphTests;
import org.jgrapht.graph.DefaultEdge;
import org.jgrapht.graph.SimpleGraph;
import org.apache.commons.math3.util.Combinations;

import java.util.*;

public class Algoritmos {

    /**
     * Contador de operaciones interno para medir pasos.
     */
    public static class OpCounter {
        private long count = 0;
        /** Incrementa en 1 */
        public void inc() { count++; }
        /** Incrementa en n */
        public void inc(long n) { count += n; }
        /** Obtiene el total */
        public long get() { return count; }
    }

    /**
     * Clona un grafo (vértices + aristas) contando operaciones.
     */
    private static <V> Graph<V, DefaultEdge> cloneGraph(
            Graph<V, DefaultEdge> g,
            OpCounter counter
    ) {
        counter.inc();
        Graph<V, DefaultEdge> copy = new SimpleGraph<>(DefaultEdge.class);
        for (V v : g.vertexSet()) {
            copy.addVertex(v);
            counter.inc();
        }
        for (DefaultEdge e : g.edgeSet()) {
            copy.addEdge(g.getEdgeSource(e), g.getEdgeTarget(e));
            counter.inc();
        }
        return copy;
    }

    /**
     * Resultado de Fuerza Bruta, con flag y conjunto de aristas.
     */
    public static class BFResult<V> {
        public final boolean found;
        public final Set<DefaultEdge> edges;
        public BFResult(boolean f, Set<DefaultEdge> e) {
            this.found = f;
            this.edges = e;
        }
    }

    /**
     * 1) Fuerza Bruta: prueba todas las combinaciones de k aristas.
     */
    public static <V> BFResult<V> bipartiteSubsetBF(
            Graph<V, DefaultEdge> G,
            int k,
            OpCounter counter
    ) {
        if (k < 0) throw new IllegalArgumentException("k debe ser ≥ 0");
        if (k > G.edgeSet().size()) throw new IllegalArgumentException("k debe ser ≤ número de aristas");

        counter = (counter != null) ? counter : new OpCounter();
        // Crear grafo auxiliar vacío con todos los vértices
        Graph<V, DefaultEdge> H = new SimpleGraph<>(DefaultEdge.class);
        counter.inc();  // Creación del grafo vacío
        for (V v : G.vertexSet()) {
            H.addVertex(v);
            counter.inc();  // Añadir cada vértice
        }

        if (k <= 2) {
            counter.inc();  // Caso trivial (0, 1 ó 2 aristas siempre forman subgrafo bipartito)
            return new BFResult<>(true, Collections.emptySet());
        }

        List<DefaultEdge> edgesList = new ArrayList<>(G.edgeSet());
        // Recorrer todas las combinaciones de k aristas
        Combinations combIterator = new Combinations(edgesList.size(), k);
        for (int[] combination : combIterator) {
            // Añadir las aristas de esta combinación al grafo H
            for (int idx : combination) {
                DefaultEdge e = edgesList.get(idx);
                H.addEdge(G.getEdgeSource(e), G.getEdgeTarget(e));
            }
            counter.inc(combination.length);  // Añadidas k aristas al subgrafo

            counter.inc();  // Operación de comprobar bipartito
            if (GraphTests.isBipartite(H)) {
                // Si encontramos un subgrafo bipartito de k aristas, devolvemos resultado
                return new BFResult<>(true, new HashSet<>(H.edgeSet()));
            }
            // Si no es bipartito, remover las aristas añadidas antes de probar la siguiente combinación
            for (int idx : combination) {
                DefaultEdge e = edgesList.get(idx);
                // Remover arista añadida (buscándola por sus extremos, ya que H tiene aristas distintas a G)
                H.removeEdge(G.getEdgeSource(e), G.getEdgeTarget(e));
            }
            counter.inc(combination.length);  // Removidas k aristas del subgrafo
        }
        // Si ninguna combinación produce un subgrafo bipartito de tamaño k
        return new BFResult<>(false, Collections.emptySet());
    }

    /**
     * Versión booleana de BF (devuelve solo flag).
     */
    public static <V> boolean bipartiteSubsetBF_bool(
            Graph<V, DefaultEdge> G,
            int k,
            OpCounter counter
    ) {
        return bipartiteSubsetBF(G, k, counter).found;
    }

    /**
     * 2) Divide & Conquer: inclusión/exclusión recursiva.
     */
    public static <V> boolean bipartiteSubsetDC(
            Graph<V, DefaultEdge> G,
            int k,
            OpCounter counter
    ) {
        if (k < 0) throw new IllegalArgumentException("k debe ser ≥ 0");
        if (k > G.edgeSet().size()) throw new IllegalArgumentException("k debe ser ≤ número de aristas");

        counter = (counter != null) ? counter : new OpCounter();
        // Grafo auxiliar H vacío (con todos los vértices de G, sin aristas)
        Graph<V, DefaultEdge> H = new SimpleGraph<>(DefaultEdge.class);
        counter.inc();  // Creación del grafo H
        for (V v : G.vertexSet()) {
            H.addVertex(v);
            counter.inc();  // Añadir cada vértice a H
        }

        if (k <= 2) {
            return true;
        }

        List<DefaultEdge> edges = new ArrayList<>(G.edgeSet());
        return dcRec(H, G, k, edges, counter);
    }

    private static <V> boolean dcRec(
            Graph<V, DefaultEdge> H,
            Graph<V, DefaultEdge> G,
            int kLeft,
            List<DefaultEdge> edges,
            OpCounter counter
    ) {
        if (kLeft == 0) {
            return true;
        }
        if (edges.size() < kLeft) {
            return false;
        }

        // Rama EXCLUIR la primera arista
        Graph<V, DefaultEdge> H_exclude = cloneGraph(H, counter);  // Copia del grafo actual
        boolean exclResult = dcRec(H_exclude, G, kLeft, edges.subList(1, edges.size()), counter);
        counter.inc();  // Operación tras completar rama "excluir"
        if (exclResult) {
            return true;
        }

        // Rama INCLUIR la primera arista
        DefaultEdge firstEdge = edges.get(0);
        Graph<V, DefaultEdge> H_include = cloneGraph(H, counter);  // Copia del grafo actual
        H_include.addEdge(G.getEdgeSource(firstEdge), G.getEdgeTarget(firstEdge));
        counter.inc();  // Añadir una arista al subgrafo
        boolean inclResult = false;
        if (GraphTests.isBipartite(H_include)) {
            inclResult = dcRec(H_include, G, kLeft - 1, edges.subList(1, edges.size()), counter);
        }
        counter.inc();  // Operación tras completar rama "incluir"
        if (inclResult) {
            return true;
        }
        // Si ninguna rama encontró solución bipartita de tamaño k, retrocede
        return false;
    }

    /**
     * 3) Backtracking: añade aristas y retrocede si no bipartito.
     */
    public static <V> boolean bipartiteSubsetBT(
            Graph<V, DefaultEdge> G,
            int k,
            OpCounter counter
    ) {
        if (k < 0) throw new IllegalArgumentException("k debe ser ≥ 0");
        if (k > G.edgeSet().size()) throw new IllegalArgumentException("k debe ser ≤ número de aristas");

        counter = (counter != null) ? counter : new OpCounter();
        if (k <= 2) {
            counter.inc();  // Caso trivial (subgrafo de 0,1,2 aristas es bipartito)
            return true;
        }

        // Grafo auxiliar H vacío (con todos los vértices de G, sin aristas)
        Graph<V, DefaultEdge> H = new SimpleGraph<>(DefaultEdge.class);
        counter.inc();  // Creación del grafo H
        for (V v : G.vertexSet()) {
            H.addVertex(v);
            counter.inc();  // Añadir cada vértice a H
        }

        List<DefaultEdge> edges = new ArrayList<>(G.edgeSet());
        // Llamada inicial a la función recursiva de backtracking
        return btRec(H, G, k, 0, edges, 0, counter);
    }

    private static <V> boolean btRec(
            Graph<V, DefaultEdge> H,
            Graph<V, DefaultEdge> G,
            int k,
            int chosen,
            List<DefaultEdge> edges,
            int start,
            OpCounter counter
    ) {
        if (chosen == k) {
            counter.inc();  // k aristas elegidas (solución completa encontrada)
            return true;
        }
        if (edges.size() - start < k - chosen) {
            counter.inc();  // No hay suficientes aristas restantes para completar k
            return false;
        }

        for (int i = start; i < edges.size(); i++) {
            DefaultEdge e = edges.get(i);
            V src = G.getEdgeSource(e);
            V tgt = G.getEdgeTarget(e);
            // Añadir la arista edges[i] al grafo H
            DefaultEdge addedEdge = H.addEdge(src, tgt);
            counter.inc();  // Añadir arista actual
            if (GraphTests.isBipartite(H)) {
                counter.inc();  // Preparar llamada recursiva
                if (btRec(H, G, k, chosen + 1, edges, i + 1, counter)) {
                    return true;
                }
            }
            // Remover la arista añadida (backtrack)
            H.removeEdge(addedEdge);
            counter.inc();  // Remover arista actual
        }
        counter.inc();  // Fin de esta rama sin encontrar solución
        return false;
    }

    /**
     * Test de bipartición utilizando JGraphT (utilidad auxiliar opcional).
     */
    private static <V> boolean isBipartite(Graph<V, DefaultEdge> g) {
        return GraphTests.isBipartite(g);
    }
}
